### DISCUSSION TITLE
Code style: `this->m_member` vs just `m_member`

### DISCUSSION THREAD

**Discussion by paulgessinger** [acts-project/acts#1008] [General]
The eternal struggle.

I notices @stephenswat recently added some code which uses the style `this->m_member` to access member variables and methods. I don't think we use this anywhere else, and rely on the prefix `m_X` to mark members relative to local variables. In a few cases we do use it to disambiguate. 

Do we want to include this in our code guidelines in some form? Encourage / discourage it?

This doesn't have to be absolutely consistent, it's just something I thought we should think about.

**Comment by HadrienG2** [comment]
Since we have lots and lots of `m_xyz` around, I'd just use `m_xyz` everywhere for consistency and minimal migration effort.

**Comment by stephenswat** [comment]
So the issue here is that using the variable without de-referencing `this` is not always possible! As I understand it, you can use the `this`-less form only for members of the class itself, not members of a parent class.

That is, unless I've really misunderstood something.

**Comment by HadrienG2** [comment]
[I'm afraid godbolt seems to disagree with you on this matter.](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:2,endLineNumber:16,positionColumn:2,positionLineNumber:16,selectionStartColumn:2,selectionStartLineNumber:16,startColumn:2,startLineNumber:16),source:'%23include+%3Ciostream%3E%0A%0Astruct+A+%7B%0A++++int+m_int+%3D+42%3B%0A%7D%3B%0A%0Astruct+B+:+A+%7B%0A++++int+method()+%7B%0A++++++++return+m_int%3B%0A++++%7D%0A%7D%3B%0A%0Aint+main()+%7B%0A++++B+b%3B%0A++++std::cout+%3C%3C+b.method()+%3C%3C+std::endl%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:50.243364514411745,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:g112,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'1',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-std%3Dc%2B%2B17',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'x86-64+gcc+11.2+(C%2B%2B,+Editor+%231,+Compiler+%231)',t:'0')),k:49.75663548558826,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,fontUsePx:'0',wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+11.2+(Compiler+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:49.75663548558826,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

**Comment by HadrienG2** [comment]
[This bit from cppreference](https://en.cppreference.com/w/cpp/language/member_functions) is interesting though:

> Within the body of a non-static member function of X, any id-expression E (e.g. an identifier) that resolves to a non-type non-static member of X or of a base class of X, is transformed to a member access expression (*this).E (unless it's already a part of a member access expression). **This does not occur in template definition context, so a name may have to be prefixed with this-> explicitly to become dependent.**

To translate the standardese into a human-readable form, C++ template parsing is hopelessly broken and because of that you may need seemingly superfluous this qualifiers in certain (obviously unspecified, otherwise it would spoil all the fun of discovering it yourself) template contexts.

**Reply by HadrienG2** [reply]
[But at least the basic stuff seems to work as expected.](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:2,endLineNumber:18,positionColumn:2,positionLineNumber:18,selectionStartColumn:2,selectionStartLineNumber:18,startColumn:2,startLineNumber:18),source:'%23include+%3Ciostream%3E%0A%0Astruct+A+%7B%0A++++int+m_int+%3D+42%3B%0A%7D%3B%0A%0Atemplate+%3Ctypename+T%3E%0Astruct+B+:+A+%7B%0A++++template+%3Ctypename+U%3E%0A++++int+method()+%7B%0A++++++++return+m_int%3B%0A++++%7D%0A%7D%3B%0A%0Aint+main()+%7B%0A++++B%3Cint%3E+b%3B%0A++++std::cout+%3C%3C+b.method%3Cint%3E()+%3C%3C+std::endl%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:50.243364514411745,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:g112,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'1',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-std%3Dc%2B%2B17',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'x86-64+gcc+11.2+(C%2B%2B,+Editor+%231,+Compiler+%231)',t:'0')),k:49.75663548558826,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,fontUsePx:'0',wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+11.2+(Compiler+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:49.75663548558826,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

**Reply by HadrienG2** [reply]
[It seems you need a templated base class and a templated child class to make it fail.](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:11,endLineNumber:16,positionColumn:11,positionLineNumber:16,selectionStartColumn:11,selectionStartLineNumber:16,startColumn:11,startLineNumber:16),source:'%23include+%3Ciostream%3E%0A%0Atemplate+%3Ctypename+T%3E%0Astruct+A+%7B%0A++++int+m_int+%3D+42%3B%0A%7D%3B%0A%0Atemplate+%3Ctypename+T%3E%0Astruct+B+:+A%3CT%3E+%7B%0A++++int+method()+%7B%0A++++++++return+m_int%3B%0A++++%7D%0A%7D%3B%0A%0Aint+main()+%7B%0A++++B%3Cint%3E+b%3B%0A++++std::cout+%3C%3C+b.method()+%3C%3C+std::endl%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:50.243364514411745,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:g112,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'0',intel:'0',libraryCode:'1',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!(),options:'-std%3Dc%2B%2B17',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'x86-64+gcc+11.2+(C%2B%2B,+Editor+%231,+Compiler+%231)',t:'0')),k:49.75663548558826,l:'4',m:50,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compiler:1,editor:1,fontScale:14,fontUsePx:'0',wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+gcc+11.2+(Compiler+%231)',t:'0')),header:(),l:'4',m:50,n:'0',o:'',s:0,t:'0')),k:49.75663548558826,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

(And in case you're wondering why C++ templates are such a mess, a significant fraction of the answer is "unconstrained specialization". When the compiler has zero ability to even tell what members a struct template will have before instantiating it, it isn't so surprising that said compiler won't behave sanely in the presence of code that refers to struct templates with unspecified parameters.)

**Comment by paulgessinger** [comment]
In that case I guess it would make the most sense to use explicit `this` in these cases where it's needed.
